### PR TITLE
connection-retries-implementation

### PR DESCRIPTION
**Pull request by tallamohan**
Allows an adapter to automatically try again when the attempt to open a new connection on the database has a transient, infrequent error. This option can be set using the `retries` configuration. Default value is 0.

resolves #

### Checklist
 - [ ] I have run this code in development and it appears to resolve the stated issue
 - [ ] This PR includes tests, or tests are not required/relevant for this PR
 - [ ] I have updated the `CHANGELOG.md` with information about my change

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -135,6 +135,25 @@ my-teradata-db-profile:
       port: <port>
 ```
 
+### Retries
+
+Allows an adapter to automatically try again when the attempt to open a new connection on the database has a transient, infrequent error. This option can be set using the `retries` configuration.
+Default value is 0.
```
Comment (SatishChGit) [inline README.md:141]: @tallamohan, I feel, it is better to mention on number times retires based on retries value.

File: dbt/adapters/teradata/connections.py [hunk 2]
```diff
@@ -175,6 +176,25 @@ def open(cls, connection):
...
+        logger.debug("host: {}",credentials.server);
+        logger.debug("retries: {}",credentials.retries);
+
+        if credentials.retries > 0:
+            def connect():
+                connection.handle = teradatasql.connect(**kwargs)
+                connection.state = 'open'
+                return connection.handle
+
+            retryable_exceptions = [teradatasql.OperationalError,]
```
Comment (SatishChGit) [inline dbt/adapters/teradata/connections.py:188]: Is comma required here?

File: dbt/adapters/teradata/connections.py [hunk 3]
```diff
@@ -175,6 +176,25 @@ def open(cls, connection):
...
+
+            return cls.retry_connection(
+                connection,
+                connect=connect,
+                logger=logger,
+                retry_limit=credentials.retries,
+                retryable_exceptions=retryable_exceptions,
```
Comment (SatishChGit) [inline dbt/adapters/teradata/connections.py:195]: Is comma required?

File: dbt/adapters/teradata/connections.py [hunk 4]
```diff
@@ -175,6 +176,25 @@ def open(cls, connection):
...
+            )
```
Comment (SatishChGit) [inline dbt/adapters/teradata/connections.py:196]: Did you try test case like network timeout by disconnecting VPN and then reconnect VPN before number of retries expire?

**Review by SatishChGit — COMMENTED** [review]
Changes are looks good. Added few minor comments. Will approve once these addressed.

File: dbt/adapters/teradata/connections.py (hunk 4 above)
Comment (tallamohan) [reply dbt/adapters/teradata/connections.py:196]: Tested this scenario. working as expected

File: dbt/adapters/teradata/connections.py (hunk 2 above)
Comment (tallamohan) [reply dbt/adapters/teradata/connections.py:188]: Not mandatory.
Removed

File: dbt/adapters/teradata/connections.py (hunk 3 above)
Comment (tallamohan) [reply dbt/adapters/teradata/connections.py:195]: Yes

File: dbt/adapters/teradata/connections.py [hunk 5]
```diff
@@ -38,6 +38,8 @@ class TeradataCredentials(Credentials):
     partition: Optional[str] = None
     sip_support: Optional[str] = None
     teradata_values: Optional[str] = None
+    retries: int = 0
+    retry_timeout: int = 1
```
Comment (tallamohan) [inline dbt/adapters/teradata/connections.py:42]: Added retry_timeout option

File: README.md [hunk 6]
```diff
@@ -135,6 +135,27 @@ my-teradata-db-profile:
       port: <port>
 ```
 
+### Retries
+
+Allows an adapter to automatically try again when the attempt to open a new connection on the database has a transient, infrequent error. This option can be set using the `retries` configuration. Default value is 0. The default wait period between connection attempts is one second. `retry_timeout` (seconds) option allows us to adjust this waiting period.
+
+If `retries` is set to 3, the adapter will try to establish a new connection three times if an error occurs.
+
```
Comment (tallamohan) [inline README.md:143]: Updated with retry_timeout option

**Comment by tallamohan** [comment]
Done the changes suggested and added the "retry_timeout" option.
Can you please re-review this pull request?

File: dbt/adapters/teradata/connections.py (hunk 2 above)
Comment (SatishChGit) [reply dbt/adapters/teradata/connections.py]: Ojay

**Review by SatishChGit — APPROVED** [review]
Looks good

**Pull request merged**